grammar.js

```diff
@@ -1,38 +1,18 @@
-// TODO rename tokens to match https://github.com/tweag/nickel/blob/master/src/parser/lexer.rs
-
-// design goal: no hidden tokens -> all non-whitespace is parsed as some token
-// this is useful for text-highlighting
-// but may be too verbose for a syntax tree?
-// -> use fields? field('fieldname', $._hidden_token)
-
-// Precedence values are taken from lalrpop grammar
-// https://github.com/tweag/nickel/blob/master/src/grammar.lalrpop
-// In lalrpop the highest precedence is 0.
-// Higher numbers imply a lower precedence. In tree-sitter however, a high
-// number implies a high precedence. To solve this issue, while keeping
-// the numbers the same, all precedence values are negated.
-const PREC = {
-  Negate: -1,       // -1
-  StrConcat: -2,    // "a" ++ "b"
-  ArrayConcat: -2,  // [1] @ [2]
-  Mult: -3,         // 1 * 2
-  Div: -3,          // 1 / 2
-  Modulo: -3,       // 1 % 2
-  Plus: -4,         // 1 + 2
-  Sub: -4,          // 1 - 2
-  BoolNot: -5,      // !true
-  Merge: -6,        // { a = "a" } & { b = "b" }
-  RightPipe: -6,    // "Hello World" |> string.split " "
-  LessThan: -7,     // 1 < 2
-  LessOrEq: -7,     // 1 <= 2
-  GreaterThan: -7,  // 1 > 2
-  GreaterOrEq: -7,  // 1 >= 2
-  Eq: -8,           // 1 == 2
-  Neq: -8,          // 1 != 2
-  BoolAnd: -9,      // true && false
-  BoolOr: -10,      // true || false
-  Arrow: -11,       // Num -> Num
-}
+// NOTE[builtin] Nickel lexes all builtin functions in lexer.rs. This is
+// possible for use, but we can also choose to parse them in the parser
+// instead. This prevents the grammar from having to be updated when a new
+// builtin function is added. Additionally, it keeps the grammar smaller. You
+// will not see rules in this grammar that match on all builtin function
+// seperately.
+
+// NOTE[typerule] In the lalrpop grammar there is a FixedType rule that is just
+// a Types rule, but with post-processing. We don't do any post-processing in
+// tree-sitter, so we just parse them as a `Types`.
+
+// NOTE[special-infix] The lalrpop grammer produces an AST. In tree-sitter we
+// don't have to do this. Hence, we don't have to treat the "|>" and "!="
+// operators differently from others. This means we can unify them with the
+// other _b_op rules.
```

Comment:
nitpick: we still could use this parser as frontend for an interpreter, by switching on `node.text`
its just that string-types are slower to compare than numeric types